src/meta/greedy_load_balancer.cpp

```diff
@@ -1116,6 +1154,105 @@ void greedy_load_balancer::get_node_migration_info(const node_state &ns,
+        ddebug_f("every app is balanced and the cluster as a whole is balanced");
+        return false;
+    }
+
+    /**
+     * Among the apps with maximum skew, attempt to pick a app where there is
+     * a move that improves the app skew and the cluster skew, if possible. If
+     * not, attempt to pick a move that improves the app skew.
+     **/
+    std::set<rpc_address> cluster_min_count_nodes;
+    std::set<rpc_address> cluster_max_count_nodes;
+    get_min_max_set(cluster_info.replicas_count, cluster_min_count_nodes, cluster_max_count_nodes);
+
+    bool found = false;
+    auto app_range = app_skew_multimap.equal_range(max_app_skew);
+    for (auto iter = app_range.first; iter != app_range.second; ++iter) {
+        auto app_id = iter->second;
+        auto it = cluster_info.apps_info.find(app_id);
+        if (it == cluster_info.apps_info.end()) {
+            continue;
+        }
+        auto app_map = it->second.replicas_count;
+        std::set<rpc_address> app_min_count_nodes;
+        std::set<rpc_address> app_max_count_nodes;
+        get_min_max_set(app_map, app_min_count_nodes, app_max_count_nodes);
+
+        /**
+         * Compute the intersection of the replica servers most loaded for the app
+         * with the replica servers most loaded overall, and likewise for least loaded.
+         * These are our ideal candidates for moving from and to, respectively.
+         **/
+        std::set<rpc_address> app_cluster_min_set =
+            utils::get_intersection(app_min_count_nodes, cluster_min_count_nodes);
+        std::set<rpc_address> app_cluster_max_set =
+            utils::get_intersection(app_max_count_nodes, cluster_max_count_nodes);
+
+        /**
+         * Do not move replicas of a balanced app if the least (most) loaded
+         * servers overall do not intersect the servers hosting the least (most)
+         * replicas of the app. Moving a replica in that case might keep the
+         * cluster skew the same or make it worse while keeping the app balanced.
+         **/
+        std::multimap<uint32_t, rpc_address> app_count_multimap = utils::flip_map(app_map);
+        if (app_count_multimap.rbegin()->first <= app_count_multimap.begin()->first + 1 &&
+            (app_cluster_min_set.empty() || app_cluster_max_set.empty())) {
+            ddebug_f(
+                "do not move replicas of a balanced app if the least (most) loaded servers overall "
+                "do not intersect the servers hosting the least (most) replicas of the app");
+            continue;
+        }
+
+        if (pick_up_move(cluster_info,
+                         app_cluster_max_set.empty() ? app_max_count_nodes : app_cluster_max_set,
+                         app_cluster_min_set.empty() ? app_min_count_nodes : app_cluster_min_set,
+                         app_id,
+                         selected_pid,
+                         next_move)) {
+            ddebug_f("found a move to make cluster more balanced");
```

Comment:
I think this log level can be deleted, otherwise there will be too many log. 
If you would like remain the log for audit, I recommend adding the app_id, pid, source machine, target machine.